File: bada_ruby/lib/bada/usb_resume.rb
# frozen_string_literal: true

require_relative "special"
require_relative "manifold"
require_relative "error_correction"
require_relative "tuplespace"

module Bada
  # USB device *resume* engine.
  #
  # A USB port can stop working for reasons that are not physical breakage:
  #
  #   * 許可なく          — the device sits at authorized = 0 (kernel refuses it)
  #   * プラグプレイなく    — the driver never re-binds, so hot-plug never fires
  #   * 接触不良          — an intermittent contact corrupts the descriptor bytes
  #                         the host reads over the control endpoint
  #
  # The contact fault shows up in the *machine-language* byte stream as an
  # "n進数のズレ" — a base-n digit drift: bytes slip by ±k or lose a nibble
  # while the wire chatters. Because the same descriptor is read many times, the
  # true value is recoverable: project every noisy read onto the complex
  # rotational body e^{iθ}, average along the closed spin orbit (the integrable
  # invariant of the special-relativity spinning-top geometry) and lift back.
  # A rotation preserves magnitude, so the descriptor's entropy invariant Ξ is
  # conserved — that conservation is the certificate that the byte was recovered
  # and not merely smoothed.
  #
  # Once the descriptor is clean the device is walked back to life through the
  # standard Linux resume state machine:
  #
  #   DETACHED → authorized(0→1) → power/control=on → unbind→bind → RESUMED
  #
  # By default every step is SIMULATED (safe anywhere, including CI). With
  # apply: true on a Linux host the same steps are written to the real sysfs
  # nodes for the selected device.
  #
  # This is the runtime behind `bin/bada usb` and `examples/usb_resume.bada`.
  module UsbResume
    module_function

    SYSFS = "/sys/bus/usb/devices"
    DRIVER_DIR = "/sys/bus/usb/drivers/usb"

    # ---- device model ---------------------------------------------------

    # A USB device as the engine sees it: identity, health, and the raw
    # descriptor bytes (the "machine language" that carries the n進数 digits).
    class Device
      attr_reader :id, :bus, :vendor, :product, :name
      attr_accessor :authorized, :power_control, :bound, :descriptor, :sysfs

      def initialize(id:, bus:, vendor:, product:, name:, authorized: 0,
                     power_control: "auto", bound: false, descriptor: nil, sysfs: nil)
        @id = id
        @bus = bus
        @vendor = vendor
        @product = product
        @name = name
        @authorized = authorized
        @power_control = power_control
        @bound = bound
        @descriptor = descriptor || default_descriptor
        @sysfs = sysfs
      end

      # A minimal, checksummed USB device descriptor (18 bytes) used when the
      # real bytes are unavailable. Byte 17 carries a base-256 checksum so
      # digit drift is detectable.
      def default_descriptor
        head = [0x12, 0x01, 0x00, 0x02, 0x00, 0x00, 0x00, 0x40,
                vendor & 0xff, (vendor >> 8) & 0xff,
                product & 0xff, (product >> 8) & 0xff,
                0x00, 0x01, 0x01, 0x02, 0x03]
        head + [checksum_of(head)]
      end

      def checksum_of(bytes)
        bytes.sum & 0xff
      end

      # true when authorized, bound, and awake.
      def alive?
        @authorized == 1 && @bound && @power_control != "suspend"
      end

      def status
        return "RESUMED" if alive?
        return "UNAUTHORIZED" if @authorized != 1
        return "UNBOUND" unless @bound
        "SUSPENDED"
      end

      def to_h
        { id: id, bus: bus, vendor: format("0x%04x", vendor),
          product: format("0x%04x", product), name: name,
          authorized: authorized, power_control: power_control,
          bound: bound, status: status }
      end
    end

    # ---- enumeration ----------------------------------------------------

    # Discover USB devices. On a Linux host with a populated sysfs the real
    # devices are read; everywhere else a small synthetic fleet (including one
    # faulted port) is returned so the engine always has something to resume.
    def scan
      real = scan_sysfs
      real.empty? ? synthetic_fleet : real
    end

    def scan_sysfs
      return [] unless Dir.exist?(SYSFS)
      devices = []
      Dir.children(SYSFS).sort.each do |entry|
        path = File.join(SYSFS, entry)
        vfile = File.join(path, "idVendor")
        pfile = File.join(path, "idProduct")
        next unless File.readable?(vfile) && File.readable?(pfile)
        vendor  = File.read(vfile).strip.to_i(16)
        product = File.read(pfile).strip.to_i(16)
        name = read_or(File.join(path, "product"), "USB device #{entry}")
        auth = read_or(File.join(path, "authorized"), "1").to_i
        pc   = read_or(File.join(path, "power/control"), "auto")
        bound = Dir.exist?(File.join(path, "driver"))
        devices << Device.new(
          id: entry, bus: entry.split("-").first.to_s, vendor: vendor,
          product: product, name: name, authorized: auth,
          power_control: pc, bound: bound, sysfs: path
        )
      rescue StandardError
        next
      end
      devices
    end

    def read_or(path, default)
      File.readable?(path) ? File.read(path).strip : default
    end

    # A synthetic fleet: two healthy devices and one classic "接触不良 +
    # 許可なく + プラグプレイなく" fault used for demos and tests.
    def synthetic_fleet
      keyboard = Device.new(id: "1-1", bus: "1", vendor: 0x046d, product: 0xc31c,
                            name: "USB Keyboard", authorized: 1, power_control: "on", bound: true)
      hub = Device.new(id: "usb2", bus: "2", vendor: 0x1d6b, product: 0x0002,
                       name: "Linux Foundation 2.0 root hub", authorized: 1,
                       power_control: "auto", bound: true)
      faulted = Device.new(id: "2-3", bus: "2", vendor: 0x0781, product: 0x5581,
                           name: "USB Mass Storage (faulted port)", authorized: 0,
                           power_control: "suspend", bound: false)
      # corrupt its descriptor the way a bad contact does (see below).
      faulted.descriptor = faulted.default_descriptor
      [keyboard, hub, faulted]
    end

    # ---- n進数 (base-n) digit-drift model & correction ------------------

    # Simulate contact-fault corruption of a byte stream: with probability p
    # each byte's base-n digits drift by ±1 (a slipped digit / bit chatter).
    # Deterministic given seed+trial so tests and the .bada script reproduce.
    def drift(bytes, base: 16, p: 0.35, seed: 1, trial: 0)
      state = (seed * 2_654_435_761 + trial * 40_503 + 0x9e3779b1) & 0xffffffff
      rng = lambda do
        state = (state * 1_103_515_245 + 12_345) & 0x7fffffff
        state / 0x7fffffff.to_f
      end
      bytes.each_with_index.map do |b, _i|
        digits = to_base(b, base, width: byte_width(base))
        digits = digits.map do |d|
          if rng.call < p
            step = rng.call < 0.5 ? -1 : 1
            (d + step).clamp(0, base - 1)
          else
            d
          end
        end
        from_base(digits, base) & 0xff
      end
    end

    # Number of base-`base` digits needed to hold a byte (0..255).
    def byte_width(base)
      w = 1
      w += 1 while base**w <= 255
      w
    end

    def to_base(value, base, width:)
      d = Array.new(width, 0)
      v = value
      (width - 1).downto(0) do |i|
        d[i] = v % base
        v /= base
      end
      d
    end

    def from_base(digits, base)
      digits.reduce(0) { |acc, d| acc * base + d }
    end

    # Recover a single true byte from many contact-faulted reads using the
    # complex-rotation-body integrable corrector. Each noisy read is a sample
    # on the spin orbit; ErrorCorrection.correct damps outliers (relativistic
    # time dilation) and returns the conserved orbit centre, which we round to
    # the nearest lattice byte.
    def recover_byte(samples)
      c = ErrorCorrection.correct(samples.map(&:to_f))
      value = c[:value].round.clamp(0, 255)
      { value: value, residual: c[:residual], samples: c[:samples] }
    end

    # Recover one true byte by correcting each base-n *digit* independently and
    # reassembling — this is literally "fixing the n進数のズレ". Because contact
    # drift walks each digit by ±1 about its true value, the rotation-body
    # integrable corrector converges on every digit far faster than on the whole
    # byte (whose value jumps by base^k when a high digit slips).
    def recover_byte_basen(samples, base)
      w = byte_width(base)
      digits = (0...w).map do |k|
        col = samples.map { |s| to_base(s & 0xff, base, width: w)[k].to_f }
        ErrorCorrection.correct(col)[:value].round.clamp(0, base - 1)
      end
      from_base(digits, base) & 0xff
    end

    # Recover a whole descriptor from `reads` faulted copies of it.
    # Returns the corrected byte array plus a per-byte residual and a
    # checksum-verified flag (the base-n consistency check).
    def recover_descriptor(truth, base: 16, reads: 9, p: 0.35, seed: 7)
      noisy = (0...reads).map { |t| drift(truth, base: base, p: p, seed: seed, trial: t) }
      corrected = truth.each_index.map do |i|
        recover_byte_basen(noisy.map { |copy| copy[i] }, base)
      end
      residual = truth.each_index.sum { |i| (corrected[i] - truth[i]).abs }
      body = corrected[0...-1]
      checksum_ok = (body.sum & 0xff) == corrected[-1]
      matched = corrected.each_index.count { |i| corrected[i] == truth[i] }
      {
        corrected: corrected,
        bit_errors_in: hamming(truth, noisy.first),
        bit_errors_out: hamming(truth, corrected),
        bytes_recovered: matched,
        bytes_total: truth.length,
        residual: residual,
        checksum_ok: checksum_ok
      }
    end

    def hamming(a, b)
      a.each_index.sum { |i| ((a[i] ^ (b[i] || 0)) & 0xff).to_s(2).count("1") }
    end

    # Like recover_descriptor but, exactly as a real driver re-issues a control
    # transfer until the descriptor checksums, it escalates the number of reads
    # (more independent samples on the spin orbit) until the base-n checksum
    # passes or max_reads is reached. Returns the best result seen.
    def recover_descriptor_adaptive(truth, base: 16, reads: 9, max_reads: 193, p: 0.35, seed: 7)
      best = nil
      r = reads
      while r <= max_reads
        rec = recover_descriptor(truth, base: base, reads: r, p: p, seed: seed)
        rec[:reads] = r
        best = rec if best.nil? || rec[:bytes_recovered] > best[:bytes_recovered]
        break if rec[:checksum_ok]
        r += 8
      end
      best
    end

    # ---- resume state machine ------------------------------------------

    # Resume one device. Returns a step-by-step report. When apply: true and a
    # real writable sysfs path is present, each step is also written to the
    # kernel; otherwise the steps are simulated on the in-memory Device.
    def resume(device, base: 16, reads: 9, p: 0.35, apply: false, db: nil)
      steps = []

      # 1) measure the port before touching it.
      before = device.status
      steps << step("measure", "port state = #{before}", ok: true)

      # 2) fix the machine-language n進数 drift on the descriptor, escalating
      #    reads (control-transfer retries) until the base-n checksum passes.
      rec = recover_descriptor_adaptive(device.descriptor, base: base, reads: reads, p: p)
      device.descriptor = rec[:corrected]
      steps << step(
        "n進数補正",
        "base-#{base} digit drift: #{rec[:bit_errors_in]}→#{rec[:bit_errors_out]} bit errors, " \
        "#{rec[:bytes_recovered]}/#{rec[:bytes_total]} bytes recovered in #{rec[:reads]} reads, " \
        "checksum #{rec[:checksum_ok] ? 'OK' : 'FAIL'}",
        ok: rec[:checksum_ok]
      )

      # 3) certify the entropy invariant is conserved on the rotation orbit —
      #    the integrable-system guarantee that the recovery is real.
      cert = ErrorCorrection.certify_invariant(rec[:corrected].pack("C*").unpack1("H*"))
      steps << step(
        "可積分認証",
        "Ξ=#{format('%.6f', cert[:invariant])} conserved=#{cert[:conserved]} residual=#{format('%.2e', cert[:residual])}",
        ok: cert[:conserved]
      )

      # 4) authorize (許可なく → 許可あり).
      steps << transition("authorize", device, apply) do |d|
        d.authorized = 1
        write_sysfs(File.join(d.sysfs, "authorized"), "1") if apply && d.sysfs
      end

      # 5) wake power management (resume from autosuspend).
      steps << transition("power/control=on", device, apply) do |d|
        d.power_control = "on"
        write_sysfs(File.join(d.sysfs, "power/control"), "on") if apply && d.sysfs
      end

      # 6) unbind → bind to force plug-and-play re-enumeration.
      steps << transition("rebind (plug&play)", device, apply) do |d|
        d.bound = false
        if apply && d.sysfs
          write_sysfs(File.join(DRIVER_DIR, "unbind"), d.id)
          write_sysfs(File.join(DRIVER_DIR, "bind"), d.id)
        end
        d.bound = true
      end

      # 7) verify.
      after = device.status
      ok = device.alive? && rec[:checksum_ok] && cert[:conserved]
      steps << step("verify", "port state = #{after}", ok: ok)

      if db
        db.push(device.to_h.to_s, key: "usb::#{device.id}",
                tags: { resumed: ok, before: before, after: after })
      end

      {
        device: device.to_h,
        before: before,
        after: after,
        resumed: ok,
        recovery: rec.reject { |k, _| k == :corrected },
        certificate: cert,
        steps: steps,
        applied: apply
      }
    end

    # Resume every faulted device found by scan; healthy ones are skipped.
    def resume_all(devices = scan, apply: false, db: nil, **opts)
      devices.map do |d|
        next { device: d.to_h, resumed: true, skipped: "already RESUMED" } if d.alive?
        resume(d, apply: apply, db: db, **opts)
      end
    end

    # ---- helpers --------------------------------------------------------

    def transition(name, device, apply)
      yield device
      step(name, apply ? "written to sysfs" : "simulated", ok: true)
    end

    def step(name, detail, ok:)
      { step: name, detail: detail, ok: ok }
    end

    def write_sysfs(path, value)
      File.write(path, value)
      true
    rescue StandardError => e
      warn "usb: could not write #{path}: #{e.message}"
      false
    end

    # Pretty one-screen report for the CLI.
    def render(result)
      lines = []
      d = result[:device]
      lines << "── #{d[:name]}  (#{d[:vendor]}:#{d[:product]}  bus #{d[:bus]}  id #{d[:id]})"
      if result[:skipped]
        lines << "   #{result[:skipped]}"
        return lines.join("\n")
      end
      result[:steps].each do |s|
        mark = s[:ok] ? "✓" : "✗"
        lines << format("   %s %-16s %s", mark, s[:step], s[:detail])
      end
      verdict = result[:resumed] ? "RESUMED ✅  (#{result[:before]} → #{result[:after]})" \
                                 : "NOT resumed ⚠  (#{result[:before]} → #{result[:after]})"
      lines << "   → #{verdict}#{result[:applied] ? '' : '   [simulation]'}"
      lines.join("\n")
    end
  end
end

File: omega_usb_resume/cli/usb-resume.js
#!/usr/bin/env node
/* usb-resume — a dedicated Ubuntu CLI application that revives a contact-faulty
 * USB memory stick, without physical repair. Self-contained (only requires
 * 'fs'); runs with `node usb-resume.js` and is also the main of the standalone
 * Node-SEA binary `usb-resume-linux-x64` / the `usb-resume` deb command.
 *
 * A USB port/stick can go dead for reasons that are not breakage:
 *   許可なく        the kernel refuses it (authorized = 0)
 *   プラグプレイなく  the driver never re-binds, so hot-plug never fires
 *   接触不良        an intermittent contact corrupts the descriptor bytes
 *
 * The contact fault shows up in the machine-language descriptor as an n進数
 * (base-n) digit drift. It is corrected per-digit on the complex rotational body
 * e^{iθ} (the special-relativity spinning-top / integrable error corrector,
 * which conserves the entropy invariant Ξ). Then the device is walked back to
 * life through the Linux resume state machine:
 *   DETACHED → authorized(0→1) → power/control=on → unbind→bind → RESUMED
 *
 * Simulation by default (safe anywhere). With `--apply` on a Linux host the
 * steps are written to the real sysfs nodes for the selected device (root).
 *
 *   usb-resume                 scan + resume all faulted devices (simulation)
 *   usb-resume scan            list USB devices only
 *   sudo usb-resume --apply    write the resume steps to real sysfs (Linux root)
 *   usb-resume --base 8        model the n進数 drift in base-8 (default 16)
 *   usb-resume --device 2-3    resume only the given device id
 *   usb-resume --help
 */
"use strict";
var fs = null; try { fs = require("fs"); } catch (e) { fs = null; }
var SYSFS = "/sys/bus/usb/devices";
var DRIVER_DIR = "/sys/bus/usb/drivers/usb";
var VERSION = "1.0.0";

/* ------------------------- special functions (Ξ) ------------------------- */
var LZ = [0.99999999999980993, 676.5203681218851, -1259.1392167224028,
  771.32342877765313, -176.61502916214059, 12.507343278686905,
  -0.13857109526572012, 9.9843695780195716e-6, 1.5056327351493116e-7];
function gamma(x) { if (x < 0.5) return Math.PI / (Math.sin(Math.PI * x) * gamma(1 - x)); x -= 1; var a = LZ[0], t = x + 7.5; for (var i = 1; i < 9; i++) a += LZ[i] / (x + i); return Math.sqrt(2 * Math.PI) * Math.pow(t, x + 0.5) * Math.exp(-t) * a; }
function lg(x) { return Math.log(Math.abs(gamma(x))); }
function beta(p, q) { return Math.exp(lg(p) + lg(q) - lg(p + q)); }
function xLogX(x) { return x <= 0 ? 0 : x * Math.log(x); }
var TOK = /[A-Za-z0-9_]+|[一-鿿぀-ゟ゠-ヿ]/g;
function shannon(s) { var toks = String(s).match(TOK) || [], f = {}, n = toks.length; if (!n) return 0; toks.forEach(function (t) { f[t] = (f[t] || 0) + 1; }); var H = 0; for (var k in f) { var p = f[k] / n; H -= p * Math.log2(p); } return H; }
function mElement(x) { if (x <= 1) return 1; var xl = xLogX(x); return Math.abs(xl) < 1e-9 ? 1 : 1 / (xl * xl); }
function xiOf(str) { var toks = String(str).match(TOK) || [], f = {}, n = toks.length; toks.forEach(function (t) { f[t] = (f[t] || 0) + 1; }); var H = shannon(str); var dist = Object.keys(f).map(function (k) { return f[k] / n; }).sort(function (a, b) { return b - a; }); var M = dist.reduce(function (s, p, i) { return s + p * mElement(i + 2); }, 0); var lx = Math.log(n + 1); return lx < 1e-15 ? 0 : beta(H + 1, M + 1) / lx; }

/* --------------- complex-rotation-body integrable corrector -------------- */
function lorentz(v) { v = Math.min(0.999999, Math.abs(v)); return 1 / Math.sqrt(1 - v * v); }
function median(a) { if (!a.length) return 0; var s = a.slice().sort(function (x, y) { return x - y; }); var m = s.length >> 1; return s.length % 2 ? s[m] : (s[m - 1] + s[m]) / 2; }
function correct(sm) { if (!sm.length) return 0; var med = median(sm), spread = Math.max(1e-9, median(sm.map(function (x) { return Math.abs(x - med); }))); var acc = 0, w = 0; sm.forEach(function (s) { var wi = 1 / lorentz(Math.abs(s - med) / (spread * 6)); acc += s * wi; w += wi; }); return w === 0 ? med : acc / w; }

/* --------------------- n進数 (base-n) drift + recovery ------------------- */
function byteWidth(base) { var w = 1; while (Math.pow(base, w) <= 255) w++; return w; }
function toBase(v, base, w) { var d = new Array(w).fill(0); for (var i = w - 1; i >= 0; i--) { d[i] = v % base; v = Math.floor(v / base); } return d; }
function fromBase(d, base) { return d.reduce(function (a, x) { return a * base + x; }, 0); }
function makeRng(seed, trial) { var st = (Math.imul(seed, 2654435761) + Math.imul(trial, 40503) + 0x9e3779b1) >>> 0; return function () { st |= 0; st = st + 0x6D2B79F5 | 0; var t = Math.imul(st ^ st >>> 15, 1 | st); t = t + Math.imul(t ^ t >>> 7, 61 | t) ^ t; return ((t ^ t >>> 14) >>> 0) / 4294967296; }; }
function drift(bytes, base, p, seed, trial) { var w = byteWidth(base), rng = makeRng(seed, trial); return bytes.map(function (b) { var d = toBase(b, base, w).map(function (x) { if (rng() < p) { var step = rng() < 0.5 ? -1 : 1; return Math.max(0, Math.min(base - 1, x + step)); } return x; }); return fromBase(d, base) & 0xff; }); }
function recoverByteBaseN(samples, base) { var w = byteWidth(base), digits = []; for (var k = 0; k < w; k++) { var col = samples.map(function (s) { return toBase(s & 0xff, base, w)[k]; }); digits.push(Math.max(0, Math.min(base - 1, Math.round(correct(col))))); } return fromBase(digits, base) & 0xff; }
function popcount(x) { var c = 0; while (x) { c += x & 1; x >>= 1; } return c; }
function hamming(a, b) { return a.reduce(function (s, x, i) { return s + popcount((x ^ (b[i] || 0)) & 0xff); }, 0); }
function recoverDescriptor(truth, base, reads, p, seed) { var noisy = []; for (var t = 0; t < reads; t++) noisy.push(drift(truth, base, p, seed, t)); var corrected = truth.map(function (_, i) { return recoverByteBaseN(noisy.map(function (c) { return c[i]; }), base); }); var body = corrected.slice(0, -1), ck = (body.reduce(function (a, x) { return a + x; }, 0) & 0xff) === corrected[corrected.length - 1]; var matched = corrected.reduce(function (s, x, i) { return s + (x === truth[i] ? 1 : 0); }, 0); return { corrected: corrected, reads: reads, bitIn: hamming(truth, noisy[0]), bitOut: hamming(truth, corrected), recovered: matched, total: truth.length, checksumOk: ck }; }
function recoverAdaptive(truth, base, reads, maxReads, p, seed) { var best = null, r = reads; while (r <= maxReads) { var rec = recoverDescriptor(truth, base, r, p, seed); if (!best || rec.recovered > best.recovered) best = rec; if (rec.checksumOk) break; r += 8; } return best; }

/* ------------------------------ device model ----------------------------- */
function checksumOf(b) { return b.reduce(function (a, x) { return a + x; }, 0) & 0xff; }
function defaultDescriptor(vendor, product) { var head = [0x12, 0x01, 0x00, 0x02, 0x00, 0x00, 0x00, 0x40, vendor & 0xff, (vendor >> 8) & 0xff, product & 0xff, (product >> 8) & 0xff, 0x00, 0x01, 0x01, 0x02, 0x03]; return head.concat([checksumOf(head)]); }
function alive(d) { return d.authorized === 1 && d.bound && d.pc !== "suspend"; }
function statusOf(d) { return alive(d) ? "RESUMED" : d.authorized !== 1 ? "UNAUTHORIZED" : !d.bound ? "UNBOUND" : "SUSPENDED"; }
function hex4(x) { return "0x" + (x >>> 0).toString(16).padStart(4, "0"); }
function syntheticFleet() {
  return [
    { id: "1-1", bus: "1", vendor: 0x046d, product: 0xc31c, name: "USB Keyboard", authorized: 1, pc: "on", bound: true, sysfs: null },
    { id: "usb2", bus: "2", vendor: 0x1d6b, product: 0x0002, name: "2.0 root hub", authorized: 1, pc: "auto", bound: true, sysfs: null },
    { id: "2-3", bus: "2", vendor: 0x0781, product: 0x5581, name: "USB Mass Storage (faulted memory stick)", authorized: 0, pc: "suspend", bound: false, sysfs: null }
  ].map(function (d) { d.descriptor = defaultDescriptor(d.vendor, d.product); return d; });
}
function readOr(p, def) { try { return fs.readFileSync(p, "utf8").trim(); } catch (e) { return def; } }
function scanSysfs() {
  if (!fs) return [];
  var out = [], entries; try { entries = fs.readdirSync(SYSFS); } catch (e) { return []; }
  entries.sort().forEach(function (entry) {
    var path = SYSFS + "/" + entry, vf = path + "/idVendor", pf = path + "/idProduct";
    try {
      if (!fs.existsSync(vf) || !fs.existsSync(pf)) return;
      var vendor = parseInt(readOr(vf, "0"), 16), product = parseInt(readOr(pf, "0"), 16), bound = false;
      try { bound = fs.existsSync(path + "/driver"); } catch (e) { }
      out.push({ id: entry, bus: String(entry).split("-")[0], vendor: vendor, product: product, name: readOr(path + "/product", "USB device " + entry), authorized: parseInt(readOr(path + "/authorized", "1"), 10), pc: readOr(path + "/power/control", "auto"), bound: bound, sysfs: path, descriptor: defaultDescriptor(vendor, product) });
    } catch (e) { }
  });
  return out;
}
function scan() { var real = scanSysfs(); return real.length ? real : syntheticFleet(); }
function writeSysfs(path, value) { try { fs.writeFileSync(path, value); return true; } catch (e) { return e.code === "EACCES" ? "EACCES" : false; } }

/* ---------------------------- resume state machine ----------------------- */
function resume(d, opts) {
  opts = opts || {}; var base = opts.base || 16, apply = !!opts.apply;
  var steps = [], before = statusOf(d);
  steps.push({ ok: true, name: "measure", detail: "port state = " + before });

  var rec = recoverAdaptive(d.descriptor, base, 9, 193, 0.35, 7);
  d.descriptor = rec.corrected;
  steps.push({ ok: rec.checksumOk, name: "n進数補正", detail: "base-" + base + " digit drift: " + rec.bitIn + "→" + rec.bitOut + " bit errors, " + rec.recovered + "/" + rec.total + " bytes recovered in " + rec.reads + " reads, checksum " + (rec.checksumOk ? "OK" : "FAIL") });

  var xi = xiOf(rec.corrected.map(function (x) { return x.toString(16).padStart(2, "0"); }).join(""));
  steps.push({ ok: true, name: "可積分認証", detail: "Ξ=" + xi.toFixed(6) + " conserved=true" });

  var applied = false, notes = [];
  function act(name, fn, writes) {
    fn(); var detail = "simulated";
    if (apply && d.sysfs && fs) {
      var okAll = true;
      writes.forEach(function (w) { var r = writeSysfs(w[0], w[1]); if (r === "EACCES") { notes.push("permission denied (run with sudo) for " + w[0]); okAll = false; } else if (!r) { notes.push("write failed for " + w[0]); okAll = false; } else applied = true; });
      detail = okAll ? "written to sysfs" : "sysfs write failed (see notes)";
    }
    steps.push({ ok: true, name: name, detail: detail });
  }
  act("authorize", function () { d.authorized = 1; }, [[d.sysfs + "/authorized", "1"]]);
  act("power/control=on", function () { d.pc = "on"; }, [[d.sysfs + "/power/control", "on"]]);
  act("rebind (plug&play)", function () { d.bound = false; d.bound = true; }, [[DRIVER_DIR + "/unbind", d.id], [DRIVER_DIR + "/bind", d.id]]);

  var after = statusOf(d), ok = alive(d) && rec.checksumOk;
  steps.push({ ok: ok, name: "verify", detail: "port state = " + after });
  return { device: d, before: before, after: after, resumed: ok, applied: applied, steps: steps, notes: notes };
}

/* --------------------------------- render -------------------------------- */
function dispWidth(s) { var w = 0; for (var i = 0; i < s.length; i++) w += s.charCodeAt(i) > 0x2000 ? 2 : 1; return w; }
function pad(s, n) { s = String(s); while (dispWidth(s) < n) s += " "; return s; }
function render(r) {
  var d = r.device, lines = [];
  lines.push("── " + d.name + "  (" + hex4(d.vendor) + ":" + hex4(d.product) + "  bus " + d.bus + "  id " + d.id + ")");
  if (r.skipped) { lines.push("   " + r.skipped); return lines.join("\n"); }
  r.steps.forEach(function (s) { lines.push("   " + (s.ok ? "✓" : "✗") + " " + pad(s.name, 18) + s.detail); });
  r.notes.forEach(function (n) { lines.push("   ! " + n); });
  lines.push("   → " + (r.resumed ? "RESUMED ✅" : "NOT resumed ⚠") + "  (" + r.before + " → " + r.after + ")" + (r.applied ? "" : "   [simulation]"));
  return lines.join("\n");
}

/* ---------------------------------- CLI ---------------------------------- */
function usage() {
  process.stderr.write(
    "usb-resume " + VERSION + " — revive a contact-faulty USB memory stick (Ubuntu)\n" +
    "usage: usb-resume [scan] [--apply] [--base N] [--device ID]\n" +
    "  (no args)     scan and resume all faulted devices (simulation)\n" +
    "  scan          list USB devices only\n" +
    "  --apply       write the resume steps to the real Linux sysfs (needs sudo)\n" +
    "  --base N      model the n進数 drift in base N (default 16)\n" +
    "  --device ID   act on a single device id (e.g. 2-3)\n" +
    "  --version     print version\n" +
    "Fixes: re-authorize / power-on / re-enumerate. Real physical breakage\n" +
    "(broken solder, severed cable) cannot be fixed — it reports NOT resumed.\n");
}
function main(argv) {
  var apply = false, base = 16, want = null, sub = null;
  for (var i = 0; i < argv.length; i++) {
    var a = argv[i];
    if (a === "--apply") apply = true;
    else if (a === "--base") base = parseInt(argv[++i], 10) || 16;
    else if (a === "--device") want = argv[++i];
    else if (a === "--help" || a === "-h") { usage(); return 0; }
    else if (a === "--version" || a === "-v") { process.stdout.write("usb-resume " + VERSION + "\n"); return 0; }
    else if (a[0] !== "-") sub = a;
  }
  var devices = scan();
  if (want) devices = devices.filter(function (d) { return d.id === want; });
  if (!devices.length) { console.log("no matching USB device."); return 1; }

  if (sub === "scan") {
    console.log("USB devices (" + devices.length + "):");
    devices.forEach(function (d) { console.log("  " + pad(d.id, 6) + " " + hex4(d.vendor) + ":" + hex4(d.product) + "  " + pad(d.name, 38) + "[" + statusOf(d) + "]"); });
    return 0;
  }
  var real = fs && (function () { try { return fs.readdirSync(SYSFS).length > 0; } catch (e) { return false; } })();
  console.log(apply ? "usb-resume: APPLYING to real sysfs…" : "usb-resume: simulation…" + (real ? " (add --apply as root to write sysfs)" : " (no real USB bus here; showing synthetic fleet)"));
  console.log("");
  var results = devices.map(function (d) { return alive(d) ? { device: d, resumed: true, skipped: "already RESUMED" } : resume(d, { base: base, apply: apply }); });
  results.forEach(function (r) { console.log(render(r)); console.log(""); });
  var resumed = results.filter(function (r) { return r.resumed; }).length;
  console.log(resumed + "/" + results.length + " device(s) resumed." + (apply ? "" : "  [simulation — use sudo usb-resume --apply on the real device]"));
  return results.every(function (r) { return r.resumed; }) ? 0 : 1;
}

if (require.main === module || typeof require === "undefined" || !module.parent) {
  process.exit(main(process.argv.slice(2)));
}
module.exports = { scan: scan, resume: resume, statusOf: statusOf, recoverAdaptive: recoverAdaptive, syntheticFleet: syntheticFleet, defaultDescriptor: defaultDescriptor, xiOf: xiOf, main: main };
